cmd/resource-migrator/main.go

```diff
@@ -0,0 +1,212 @@
+	c := make(chan uploadRequest, (modelListSize+worldListSize)*10)
+
+	// Requesting all models
+	log.Println("Processing Models")
+	requestUpload[*models.Model](c, ms, "models")
+
+	// Requesting all worlds
+	log.Println("Processing Worlds")
+	requestUpload[*worlds.World](c, ws, "worlds")
+
+	// Listen for exit signal from when all resources have been uploaded
+	exit := make(chan struct{}, 1)
+
+	// Begin parallel uploads and keep track of them using the progress bar. It will send a single item to the exit
+	// channel once it finished
+	bar := newProgressBar(modelListSize+worldListSize, exit)
+	for i := 0; i < maxParallelUploads; i++ {
+		go upload(c, s, bar)
+	}
+
+	// Listen for Interrupt and Terminate signals
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	for {
+		select {
+		case sig := <-sigs:
+			log.Panicln("Signal received:", sig.String())
+		case <-exit:
+			log.Println("Models and Worlds were successfully migrated. Took:", time.Since(started).Seconds(), "seconds")
+			return
+		}
+	}
+}
+
+func setup() (storage.Storage, *gorm.DB) {
+	db, err := setupDB()
+	if err != nil {
+		log.Fatalln("Failed to set up to MySQL database conn:", err)
+	}
+
+	globals.ResourceDir = os.Getenv("FUEL_RESOURCE_DIR")
+
+	// Set up git
+	globals.VCSRepoFactory = func(ctx context.Context, dirpath string) vcs.VCS {
+		return vcs.GoGitVCS{}.NewRepo(dirpath)
+	}
+
+	// Initialize S3 config
+	s3session := session.Must(session.NewSession())
+	s := storage.NewS3v1(s3.New(s3session), s3manager.NewUploader(s3session), os.Getenv("AWS_S3_BUCKET"))
+	return s, db
+}
+
+func newProgressBar(size int, exit chan struct{}) *progressbar.ProgressBar {
+	return progressbar.NewOptions(size,
+		progressbar.OptionSetPredictTime(false),
+		progressbar.OptionShowCount(),
+		progressbar.OptionSetDescription("Uploading resources"),
+		progressbar.OptionSetWriter(os.Stderr),
+		progressbar.OptionSetWidth(10),
+		progressbar.OptionThrottle(65*time.Millisecond),
+		progressbar.OptionOnCompletion(func() {
+			exit <- struct{}{}
+		}),
+		progressbar.OptionSpinnerType(14),
+		progressbar.OptionFullWidth(),
+		progressbar.OptionSetRenderBlankState(true),
+	)
+}
+
+func setupDB() (*gorm.DB, error) {
+	// Initialize database
+	cfg, err := gz.NewDatabaseConfigFromEnvVars()
+	if err != nil {
+		return nil, err
+	}
+	db, err := gz.InitDbWithCfg(&cfg)
+	if err != nil {
+		return nil, err
+	}
+	return db, err
+}
+
+type uploadRequest struct {
+	Kind     string
+	Resource res.Resource
+}
+
+func requestUpload[T res.Resource](c chan uploadRequest, items []T, kind string) {
+	for _, item := range items {
+		c <- uploadRequest{
+			Kind:     kind,
+			Resource: item,
+		}
+	}
+}
+
+func upload(c chan uploadRequest, storage storage.Storage, bar *progressbar.ProgressBar) {
+	for !bar.IsFinished() {
+		req := <-c
+		err := uploadResources(context.Background(), storage, req.Kind, req.Resource)
+		if err != nil {
+			continue
+		}
+		_ = bar.Add(1)
```

Comment:
Since the termination logic is tied to the progressbar finishing, not calling `bar.Add(1)` if there is an error will prevent the program from exiting.

A quick suggestion would be to always call `bar.Add(1)`, even before checking for errors, but if we do so we should either print the resource that failed to upload or have a separate mechanism to report models that failed to upload.

```suggestion
		_ = bar.Add(1)
		if err != nil {
			log.Printf("Failed to upload %s. Error: %s", req.Resource.GetName(), err)
		}
```

Reply:
I gave this another visit and ended up adding a channel that allow us to report errors when uploading resources. The list is later printed when the migration is over.